Add benchmark generator

kevinjalbert opened the pull request:
### Summary

Introduce benchmark generator to easily add benchmarks to Rails
applications.

CC @gmcgibbon @rafaelfranca

kevinjalbert committed:
Add benchmark generator

Introduce benchmark generator to add benchmarks to Rails applications.

The generator makes use of `benchmark-ips`, and will automatically add
the gem to the Gemfile as needed.

Co-authored-by: Gannon McGibbon <gannon.mcgibbon@gmail.com>

gmcgibbon commented on railties/lib/rails/generators/rails/benchmark/benchmark_generator.rb:
```diff
@@ -0,0 +1,30 @@
+      def generate_layout
+
+        if benchmark_method == "ips"
+          # TODO: Avoid having this run each time
```
☝️ We should analyze the gemfile to see if IPS is already added, otherwise run this code and then add a test for it.

kaspth commented on railties/lib/rails/generators/rails/benchmark/USAGE:
```diff
@@ -0,0 +1,15 @@
+
+    You can specify a different benchmarking method with `--method`.
+    Valid methods are ips, bm, and bmbm.
+    ​
```
Should mention how to run a generated benchmark too.

kaspth commented on railties/lib/rails/generators/rails/benchmark/benchmark_generator.rb:
```diff
@@ -0,0 +1,30 @@
+
+        if benchmark_method == "ips"
+          # TODO: Avoid having this run each time
+          gem_group(:development) do
```
Why not test too? Presumably you could want to set up a benchmark CI?

kevinjalbert commented:
I've addressed the comments, let me know if there is anything else.

I did notice that CI is looking a bit off (some failures that don't seem related). I might give it a rebase off master later to see if that fixes things.

kaspth commented on railties/test/generators/benchmark_generator_test.rb:
```diff
@@ -0,0 +1,118 @@
+
+        assert_file("Gemfile") do |content|
+          occurrences = content.scan("gem 'benchmark-ips'").count
+          assert_equal occurrences, 1, "Should only have benchmark-ips present once"
```
The order is off: it's expected (1) then actual (occurrences).

kaspth commented on railties/lib/rails/generators/rails/benchmark/benchmark_generator.rb:
```diff
@@ -0,0 +1,37 @@
+
+      def ips_installed?
+        in_root do
+          return File.read("Gemfile").include?("gem 'benchmark-ips'")
```
Do we need the `return` here?

kevinjalbert replied:
The `return` is needed as we're in the block scope of `in_root`. If I remove it the tests fail.

gmcgibbon replied:
[`in_root`](https://github.com/erikhuda/thor/blob/v1.0.0/lib/thor/actions.rb#L196) doesn't appear to pay attention to the return of the yield. This assumes the quoting of the gemfile. Can we do a regex like https://github.com/rails/rails/blob/d8b86e4b746962aa3b4c16cc447d9b48160c39f6/railties/lib/rails/generators/rails/db/system/change/change_generator.rb#L53 instead?

kevinjalbert replied:
@gmcgibbon I addressed this concern in a new commit.

👍 Good catch on the assumption I had with the gemfile's quoting preference.

kaspth commented on railties/lib/rails/generators/rails/benchmark/benchmark_generator.rb:
```diff
@@ -0,0 +1,37 @@
+module Rails
+  module Generators
+    class BenchmarkGenerator < NamedBase
+      argument :reports, type: :array, default: []
```
Let's find a simple clearly named 2-value default here for the common case AB-benchmark, such that people don't have to pass those. Maybe "before", "after"?

kaspth commented on railties/lib/rails/generators/rails/benchmark/templates/benchmark.rb.tt:
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+require_relative "../config/environment"
+
```
In my experience, benchmarks tend to do some setup here that they act on in the reports. Wonder if we should at least note that with a comment. Or maybe there's a way to set up some sample data that's not annoying ala `DATA = 100.times.map { "example" }`. Or:

```ruby
# Override the method under test and optimize it.
class Post
  def optimized_title
    title
  end
end

# Find a representative sample.
post = Post.first

Benchmark.ips do |x|
  x.report("before") { post.title }
  x.report("after")   { post.optimized_title }
end
```

Hell, we could consider adding a `require "active_record/test_case/benchmark_helper` then if people ran this in the `test` environment, we'd automatically wrap the script in a transaction that would be rolled back after the benchmark. Just some more ideas for what we could do since we're automating things here.

kevinjalbert replied:
🤔 I'm good with leaving a comment in the template and leave the decision with the end user.

Unless there is a general setup that works best for _setting_ things up?

The benchmark_helper is an interesting idea, although is it maybe too prescriptive?

-----

I'm 100% open to suggestions on how to move forward with this comment.

kaspth replied:
> Unless there is a general setup that works best for _setting_ things up?

That's benchmark dependent, I guess. But it's stuff like this that would be good to think about such that we can make calls on it. What would people be testing here? Are these one-off scripts and what's under test? Active Record interactions? Full view tests? (Probably not). Seems like this is at the model level, so I think it's fine to make model related decisions, while still leaving it possible to remove that stuff for other testing.

nateberkopec replied:
I like this idea.

Common benchmark mistake:

```ruby 
x.report("thing") { MyService.perform(User.first) }
```

Reifying a "setup method" or "setup area" where data is prepared sounds good to me.

gmcgibbon replied:
I assume we would do this with a comment and some whitespace above the benchmark?

kaspth commented on railties/lib/rails/generators/rails/benchmark/benchmark_generator.rb:
```diff
@@ -0,0 +1,37 @@
+  module Generators
+    class BenchmarkGenerator < NamedBase
+      argument :reports, type: :array, default: []
+      class_option :method, type: :string, enum: %w(ips bm bmbm), default: "ips"
```
How often do people actually use bm or bmbm? I thought everybody was just ips from now on?

kevinjalbert replied:
That is a good question. I personally have always used `ips` myself.

Is there value in keeping the different options present? The default is set to `ips` though.

gmcgibbon replied:
I like having the option of including `bm` and `bmbm` for cases where you only want to measure something once. That said I can't think of a common use case where that's really necessary, so maybe we can 🔥it and cut down the logic in the generator.

kevinjalbert replied:
I'm all for simplifying things. If we're good with this I'll pull out `bm` and `bmbm` from the generator.

gmcgibbon replied:
Sure, let's 🔥 it for now. We can always add it back later. That likely means `class_option :method` and all tests and docs associated with it can simply be removed.

kaspth replied:
Agreed. Let's ship something more opinionated that steers people in a certain direction initially. And reap the implementation rewards of less scope for v1.

gmcgibbon commented on railties/lib/rails/generators/rails/benchmark/USAGE:
```diff
@@ -0,0 +1,17 @@
+  method = ips
+
+Example:
+    `rails generate benchmark opt_compare before after`
```
`before` and `after` isn't strictly necessary here since it is the default.

gmcgibbon commented on railties/lib/rails/generators/rails/benchmark/benchmark_generator.rb:
```diff
@@ -0,0 +1,41 @@
+      end
+
+      def ips_gem_name
+        "benchmark-ips"
```
This can be a constant, or included in the regex as a constant.

kaspth commented on railties/lib/rails/generators/rails/benchmark/benchmark_generator.rb:
```diff
@@ -0,0 +1,41 @@
+
+      def ips_installed?
+        in_root do
+          return File.read("Gemfile").match?(/gem.*\b#{ips_gem_name}\b.*/)
```
Pretty surprised we don't already have a matcher for stuff like this, yet we have to read the Gemfile manually and custom write a regex? Seems odd. Does `gem` include something like this?

kaspth replied:
Nope. Looks like @gmcgibbon had to add this for the change generator too: https://github.com/rails/rails/blob/f619ac91d33bd431d94726cc97092c3dbeb2afc6/railties/lib/rails/generators/rails/db/system/change/change_generator.rb#L53

Let's leave this for now, and potentially look at an extraction later.

kevinjalbert replied:
>  Let's leave this for now and potentially look at an extraction later.

@kaspth just for clarification, do you mean in a separate PR afterwards?

kaspth commented on railties/lib/rails/generators/rails/benchmark/benchmark_generator.rb:
```diff
@@ -0,0 +1,41 @@
+
+      def add_ips_to_gemfile
+        gem_group(:development, :test) do
+          gem(ips_gem_name)
```
Think we should just append the group here ala `gem "benchmark-ips", group: %i[ development test ]`. It's not like we're automatically inserting the ips gem in an existing dev/test group.

kaspth commented:
Honestly, this seems lacking in terms of what we want to help with, we've rejected adding simple generators in the past (one for concerns was too little, model generators are above the threshold). If all this was doing was just generating the benchmark file, I'd reject this. Think we have to find some sum-greater-than-parts usefulness here to make this worth it. The automatic gem file entry is a sample of that.

It's rare that I write these types of benchmarks and I'd typically rather profile with production workloads via APM tools.

@nateberkopec do you run benchmarks like this for clients? Got any suggestions for where the utility of a benchmark generator for a Rails app would be?

gmcgibbon commented on railties/lib/rails/generators/rails/app/templates/Gemfile.tt:
```diff
@@ -3,7 +3,7 @@ git_source(:github) { |repo| "https://github.com/#{repo}.git" }
 ruby <%= "'#{RUBY_VERSION}'" -%>
 
-<% unless gemfile_entries.first.comment -%>
+<% unless gemfile_entries.first&.comment -%>
```
If you specify you want an empty gemfile, this assumes there's always one entry and crashes.

gmcgibbon commented:
At Shopify, we use this generator to measure optimizations to small units of code (eg. methods). However, I'd agree profiling is definitely a more popular approach. My workflow for identifying performance problems is to use profiling to identify slow code and then benchmarking to validate an optimization. 

Adversely, we don't provide a generator for profiling, and it could be argued benchmarking isn't a specific concern of web applications. I think most of the usefulness of this generator lies in having your entire app available to you in a standalone benchmarking script, and the gemfile addition.

nateberkopec commented:
> @nateberkopec do you run benchmarks like this for clients? Got any suggestions for where the utility of a benchmark generator for a Rails app would be?

I do. Usually it's a unit of logic/transaction script, e.g. `MyJob.perform` or `CustomerProcessor.do_thing`. 

Here are a few issues with Rails and benchmarking that I think a generator could solve:

1. In every single one of my workshops, we go over how to use `rails runner` to set up benchmark/ips. This takes about 15 minutes and blows everyone's minds. "How do I benchmark User.do_the_thing in Rails?" currently does not have an easily-discoverable answer, whereas "rails g benchmark" solves it all in 10 seconds.
2. There is no agreement as to where to put these benchmarks, so a generator helps set that expectation of "where benchmarks go in a Rails app".
3. I'm always surprised at my workshops how everyone has used the Benchmark stuff from core, but not benchmark/ips, which is so much better and methodologically/statistically sound.

nateberkopec commented on railties/lib/rails/generators/rails/benchmark/USAGE:
```diff
@@ -0,0 +1,17 @@
+Description:
+    Generate benchmarks to compare performance optimizations.
```
For me, the big thing is the simple significance test, automatic warmup, and the fact that you don't need to choose a # of iterations.

nateberkopec commented on railties/lib/rails/generators/rails/benchmark/templates/benchmark.rb.tt:
```diff
@@ -0,0 +1,9 @@
+Benchmark.<%= benchmark_method %> do |x|
+<%- reports.each do |report| -%>
+  x.report("<%= report %>") { }
+<%- end -%>
```
If we're gonna ship with ips default, we should `x.compare!` here

kaspth commented:
Cool, perfect. Having knowing and preferred benckmark-ips for years, I didn't realize that it wasn't well known. So this generator is a big help and I'm back to being excited about this, thanks 🙌

Perhaps we should run `bundle` automatically too after adding the gem?

As for the directory I'm not sure about having a top-level benchmarks directory. We'll end benchmarking the stuff in app/ or lib/, I think script/benchmarks would be more appropriate.

kevinjalbert commented:
I've updated the PR with some of the suggestions mentioned in the comments.

I think the last big things to figure/work out are:

- Creating a "setup method" or "setup area" where data is prepared before actually benchmarking [[comment]](https://github.com/rails/rails/pull/37948#discussion_r365104479)
- To move the generated template files under `/script/benchmarks` [[comment]](https://github.com/rails/rails/pull/37948#issuecomment-573027499)
- If we want to run bundle automatically after installing the ips gem [[comment]](https://github.com/rails/rails/pull/37948#issuecomment-573027499)
- Final copy changes with the USAGE and CHANGELOG files
- Resolve conflicts in CHANGELOG

I'm a bit unsure of the setup area/method as I'm not too familiar with these in practice. Some guidance here would be appreciated.

kevinjalbert commented:
I've just addressed these points from my last comment.

- Creating a "setup method" or "setup area" where data is prepared before actually benchmarking
- To move the generated template files under /script/benchmarks

@kaspth @gmcgibbon if you have any more comments or actions to move this pull request along?

gmcgibbon reviewed (commented):
Please squash and rebase your commits

rafaelfranca commented on railties/CHANGELOG.md:
```diff
@@ -1,3 +1,17 @@
+    By default, two code blocks can be benchmarked using the `before` and `after` reports.
+
+    You can run the generated benchmark file using:
+      `bundle exec ruby script/benchmarks/opt_compare.rb`
```
We don't need `bundle exec`. `config/environment` setup bundler so we should remove this.

rafaelfranca commented on railties/lib/rails/generators/rails/benchmark/USAGE:
```diff
@@ -0,0 +1,19 @@
+        script/benchmarks/opt_compare.rb
+
+    You can run the generated benchmark file using:
+        `bundle exec ruby script/benchmarks/opt_compare.rb`
```
Same about `bundle exec`

rafaelfranca commented on railties/lib/rails/generators/rails/benchmark/benchmark_generator.rb:
```diff
@@ -0,0 +1,29 @@
+
+      def ips_installed?
+        in_root do
+          return File.read("Gemfile").match?(/gem.*\b#{IPS_GEM_NAME}\b.*/)
```
Should we use `bundle show`?

gmcgibbon replied:
We could, I wonder which approach is faster? Probably the current one, but shelling out to `bundle show gem_name` is likely much less error prone.

Perhaps we could keep this approach now and refactor this in both the benchmark and db system generators.

kevinjalbert commented:
@rafaelfranca I've removed the instances of `bundle exec` that you mentioned.

As per the other comment about using `bundle show`. I'm good with either approach. Like @gmcgibbon mentioned, might be less error prone, but I'm also okay with making a follow-up PR that addresses both this generator and the db system generator as well.
